### ISSUE TITLE
jira server (local) import

### ISSUE THREAD

**Issue by jvl1v5** [JetBrains/space-issues-import#11]
Hi,

I'm trying to import form a local jira server and getting a 403 forbbiden status code.
The user defined in the script has admin permissions and is projectleader of the project to be imported.
The token is a "personal access token" generated in the users profile.
In the first attempts it gave me a 401 unauthorized, so I think the credentials are ok.

Errors from script
```
2023-01-06 12:41:14.125 [main] INFO  c.j.space.import.CommandLineArgs - Running in debug mode
2023-01-06 12:41:15.246 [main] ERROR c.j.s.import.jira.JiraIssuesLoader - failed to retrieve issues from Jira
2023-01-06 12:41:15.247 [main] ERROR c.j.s.import.jira.JiraIssuesLoader - original error message: RestClientException{statusCode=Optional.of(403), errorCollections=[]}
2023-01-06 12:41:15.247 [main] ERROR c.j.s.import.jira.JiraIssuesLoader - if the problem still persists, report it here: https://github.com/JetBrains/space-issues-import/issues/new
2023-01-06 12:41:15.250 [main] ERROR c.j.space.import.CommandLineArgs - Failed to load issues from external system
```

Script call (personal information replaced)
```
docker run public.registry.jetbrains.space/p/space/containers/space-issues-import:latest \
                --importSource Jira \
                --spaceServer http://space.xxxx.local \
                --spaceToken xxxx
                --spaceProject key::TEST \
                --jiraServer http://jira.xxxx.local:8080 \
                --jiraUser xxxx \
                --jiraApiToken xxxx \
                --jiraQuery "project = TEST" \
                -a "Firstname Lastname::Firstname Lastname" \
                -s "Selected for Development::Selected for Dev" \
                -s "Next::Next" \
                -s "In Arbeit::In progress" \
                -s "Fertig::Done" \
                --updateExistingIssues \
                --replaceMissingStatus \
                --replaceMissingAssignee \
                --dryRun
                --debug
```

Thank you for helping.

**Comment by dimart** [comment]
Hi @jvl1v5,

thank you for reporting it, I'd ask you to provide a bit more details:
- do you use a local self-managed Space instance (Space On-Premises)?
- could you please run the script passing `https://...` instead of `http://` for the server options?

Thanks!

**Comment by hakito** [comment]
Facing exactly the same problem.

* space on premise
* jira on premise
* https url with valid (public) certificate

I tried it 
* with powershell and with command prompt. 
* I also tried single and double quoting the jiraApiToken. 
* Tried it with the sysadmin user

The jira page shows in the column _Last authenticated_ the value _Never_.

I verified that the token is valid by invoking a curl request on the api manually:

```
curl --request GET \
  --url 'https://jira.example.com/rest/api/2/issue/DEV-1' \
  --header 'Authorization: Bearer PERSONAL_ACCESS_TOKEN' \
  --header 'Accept: application/json'
  ```
  
So what is the _--jiraUser_ required for? 
  
And a follow up question (when the import starts working): How do I tell the import which jira project to import?

**Comment by dimart** [comment]
Hi @hakito,

thank you for providing the details, I reproduced the error on my local Jira Server (Jira Software 9.6.0).

**TL;DR:** For Jira On-Premises you need to pass `--jiraUser` and `--jiraPassword`. To tell the import script which JIRA project to import from, you can pass [JQL query](https://support.atlassian.com/jira-service-management-cloud/docs/use-advanced-search-with-jira-query-language-jql/) as `--jiraQuery "project = YOUR_PROJECT_KEY"`

---

The library we use to communicate with JIRA via REST API at the moment is [JIRA REST Java Client Implementation](https://mvnrepository.com/artifact/com.atlassian.jira/jira-rest-java-client-core). It only supports `Anonymous` or `Basic Http` authentication.

For Jira Cloud, it means that you can safely pass [an API Token](https://support.atlassian.com/atlassian-account/docs/manage-api-tokens-for-your-atlassian-account/) via `--jiraApiToken` option (together with `--jiraUser`).

For Jira Server, it means that you need to pass a password since API Tokens sadly are [not available](https://community.atlassian.com/t5/Jira-questions/API-Tokens-for-self-hosted-Jira/qaq-p/820644) on JIRA On-Premises. 
The library will [encode](https://bitbucket.org/atlassian/jira-rest-java-client/src/a90e1e52b4ee21ce6a17615aa41e97d87a3d76d7/core/src/main/java/com/atlassian/jira/rest/client/auth/BasicHttpAuthenticationHandler.java?at=master#lines-47) username and password in Base64 according to the [docs](https://developer.atlassian.com/server/jira/platform/basic-authentication/#construct-the-authorization-header).

<img width="933" alt="Screenshot 2023-02-24 at 15 25 53" src="https://user-images.githubusercontent.com/2109133/221203350-f97527b3-b7e4-4165-8e71-ee9fd5816bdf.png">

\* Keep an eye on `Current failed logins` number. I noticed that when I failed to enter correct credentials a few times during my experiments, then even correct creds would not be accepted. It is due to [CAPTCHA](https://developer.atlassian.com/server/jira/platform/basic-authentication/#captcha). Logging-in via a browser and entering it helps.


Regarding your question about the way to specify a project to import from. You can pass `--jiraQuery` which accepts a valid [JQL query](https://support.atlassian.com/jira-service-management-cloud/docs/use-advanced-search-with-jira-query-language-jql/). For example, if I want to import from the project with the key `TEST`, I need to pass `"project = TEST"` for `--jiraQuery`.

Thank you for the patience, I hope it solves the problem for you.

**Comment by dimart** [comment]
@hakito @jvl1v5 if you have any other issues with running the script, feel free to reopen this one / create a new one, thanks!
